flink/v1.20/flink/src/main/java/org/apache/iceberg/flink/sink/IcebergSinkBuilder.java

```diff
@@ -0,0 +1,122 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.sink;
+
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import org.apache.flink.annotation.Internal;
+import org.apache.flink.configuration.ReadableConfig;
+import org.apache.flink.streaming.api.datastream.DataStream;
+import org.apache.flink.streaming.api.datastream.DataStreamSink;
+import org.apache.flink.table.api.TableSchema;
+import org.apache.flink.table.data.RowData;
+import org.apache.flink.types.Row;
+import org.apache.iceberg.DistributionMode;
+import org.apache.iceberg.SerializableTable;
+import org.apache.iceberg.Table;
+import org.apache.iceberg.flink.TableLoader;
+import org.apache.iceberg.relocated.com.google.common.annotations.VisibleForTesting;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.relocated.com.google.common.collect.Sets;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+@Internal
+/*
+ This class is for internal purpose of transition between the previous implementation of Flink's sink (FlinkSink)
+ and the new one implementation based on Flink v2 sink's API (IcebergSink)
+*/
```

Comment:
I'm not entirely convinced that this is the path we should move forward to. One of the design choices during the implementation of the new [IceberSink](https://github.com/apache/iceberg/pull/10179) was exactly to avoid this: instead of refactoring common code, we would just add a brand new Sink and keep the old one untouched (or minimally touched). 

In my mind, this PR is about adding additional tests using the new sink, in places where only the old one is used. If we want to keep the design choice from my previous PR, then that would imply that we would be creating new test cases altogether, but only targeting the IcebergSink. 
For example, we have
`TestFlinkIcebergSinkExtended`

then we could add a new one called
`TestFlinkIcebergSinkExtendedSinkV2`
which is identical as the original one, but replacing the `FlinkSink.` with `IcebergSink.` 

And even though this will require creating extra files, I think it will be worth it as we can keep the 2 sinks (implementations and unit testing) separate.

Wdyt? @pvary @stevenzwu

Reply:
Here are the priorities in my head:
- When removing the FlinkSink from the Iceberg connector codebase, the users of the IcebergSink should not need to change any code.
- It should be as easy as possible to remove the FlinkSink and related classes. There should not be any dangling classes left (unused interfaces and such)
- We should reuse as much of the testing code as possible

From my perspective, it's ok to add an interface to separate the common methods, but I prefer not to have any implementation in the interface, as this class should be removed when the deprecation happens. If this approach is not enough, then we need test only util methods to create the sinks for the tests

Reply:
Thank you for the review folks. The listed priorities sounds reasonable. Would you agree @pvary that we should add two more bullets to this lists:
- We should ensure that `IcebergSink` is covered by unit tests to a similar level as `FlinkSink`
- We should ensure as transitive migration from `FlinkSink` to `IcebergSink` as possible. For example, try to keep features and allowed properties synchronized (of course, I'm not talking about features that are not possible to synchronize because of Flink's API differences)
?

With regards to this: 
> When removing the FlinkSink from the Iceberg connector codebase, the users of the IcebergSink should not need to change any code.

I think that the suggested by you approach with the interface without any implementation shouldn't break the user's code. I can do a manual test simulating the removal of the interface and verify if no exception is thrown.

@rodmeneses with regards to this comment:
> I think we should follow a similar approach as the one @pvary mentioned before: defining a config, and relaying on it. For example:
> adding the following to FlinkConfigOptions

I wanted to do this, but I thought that it would be better to move this to the separated PR as I mentioned in this comment: https://github.com/apache/iceberg/pull/11219#discussion_r1781169361 WDYAT?

I started with a lower level unit tests because I saw that they cover a lot of use cases and as a potential user of this class, I would sleep better if I knew that all tests passes also for `IcebergSink`. For the purpose of this feature flag, I think that we should add a higher level tests that are using Flink's table API/Flink SQL. Do you agree on that?

Reply:
I changed abstract class to the interface.
BTW, I've just realized more about what's going on around the current status of unit tests for these two classes. I see that we already had some tests of `IcebergSink` that are copy-pastes of `FlinkSink` tests (e.g. `TestIcebergSink` vs `TestFlinkIcebergSink`). I see that these tests are already desynchronized. Don't you think that this state is becoming hard to maintain for a longer period of time?

Reply:
Hi @arkadius 
I'm OK with adding a new interface `IcerbergSinkBuilder` and having the `FlinkSink` and `IcebergSink` implement it. So on my mind, the only things that are changing in this PR are exactly those ones. I can see that in the PR already, with only one thing I noticed which I will comment directly in the PR itself.

Reply:
> I changed abstract class to the interface. BTW, I've just realized more about what's going on around the current status of unit tests for these two classes. I see that we already had some tests of `IcebergSink` that are copy-pastes of `FlinkSink` tests (e.g. `TestIcebergSink` vs `TestFlinkIcebergSink`). I see that these tests are already desynchronized. Don't you think that this state is becoming hard to maintain for a longer period of time?

During the implementation of the `IcerbergSink` I wrote unit test to cover all the new functionality specific to the new sink, and I also added unit test to cover _similar_ functionality from the `FlinkSink`. I didn't want to modify existing code (which was a design choice in this PR) so I went ahead and created new test classes that will only feature the new `IcebergSink`. This is the reason I suggest yesterday about adding new test classes that only target the new `IcbergSink`, so that it is similar to the approach that I followed.

Having said that, I really liked the idea of your new marker interface as it minimally change behavior or APIs in the current code, and at the same time it makes very simple and clear adding a parameter "useSinkV2" for the existing unit tests, so that you can use either `FlinkSink` or `IcebergSink`.